pkg/profiler/profile_writer.go

```diff
@@ -57,22 +58,30 @@ func (fw *FileProfileWriter) Write(_ context.Context, labels model.LabelSet, pro
 
 // RemoteProfileWriter is a profile writer that writes profiles to a remote profile store.
 type RemoteProfileWriter struct {
+	zw                 *gzip.Writer
 	profileStoreClient profilestorepb.ProfileStoreServiceClient
 }
 
 // NewRemoteProfileWriter creates a new RemoteProfileWriter.
 func NewRemoteProfileWriter(profileStoreClient profilestorepb.ProfileStoreServiceClient) *RemoteProfileWriter {
 	return &RemoteProfileWriter{
+		zw:                 gzip.NewWriter(nil),
 		profileStoreClient: profileStoreClient,
 	}
 }
 
 // Write sends the profile using the designated write client.
 func (rw *RemoteProfileWriter) Write(ctx context.Context, labels model.LabelSet, prof *profile.Profile) error {
+	// Write a profile into buf with a reusable gzip writer
+	// to reduce memory allocations (reuse a compressor).
+	// Note, an archive won't have a gzip footer (it still works)
+	// because we can't call Close.
 	buf := bytes.NewBuffer(nil)
-	if err := prof.Write(buf); err != nil {
+	rw.zw.Reset(buf)
+	if err := prof.WriteUncompressed(rw.zw); err != nil {
```

Comment:
I am not super confident about this solution, but it seems `gzip.Writer` reusing reduced the allocs.
Note, compress package no longer shows up in the top.

<details>

```
(pprof) top20 -cum
Showing nodes accounting for 285.37MB, 49.90% of 571.93MB total
Dropped 164 nodes (cum <= 2.86MB)
Showing top 20 nodes out of 117
      flat  flat%   sum%        cum   cum%
         0     0%     0%   211.29MB 36.94%  github.com/godbus/dbus/v5.(*Conn).inWorker
   24.12MB  4.22%  4.22%   211.29MB 36.94%  github.com/godbus/dbus/v5.(*unixTransport).ReadMessage
   31.59MB  5.52%  9.74%   186.68MB 32.64%  github.com/godbus/dbus/v5.DecodeMessage
         0     0%  9.74%   185.91MB 32.51%  github.com/oklog/run.(*Group).Run.func1
         0     0%  9.74%   185.91MB 32.51%  runtime/pprof.Do
         0     0%  9.74%   185.41MB 32.42%  github.com/parca-dev/parca-agent/pkg/profiler/cpu.(*CPU).Run
         0     0%  9.74%   185.41MB 32.42%  main.run.func6
         0     0%  9.74%   185.41MB 32.42%  main.run.func6.1
         0     0%  9.74%   155.09MB 27.12%  github.com/godbus/dbus/v5.(*decoder).Decode
  127.51MB 22.30% 32.04%   155.09MB 27.12%  github.com/godbus/dbus/v5.(*decoder).decode
    5.53MB  0.97% 33.00%   118.18MB 20.66%  github.com/coreos/go-systemd/v22/dbus.(*Conn).SubscribeUnitsCustom.func1
         0     0% 33.00%   113.51MB 19.85%  debug/elf.Open
         0     0% 33.00%   112.65MB 19.70%  github.com/coreos/go-systemd/v22/dbus.(*Conn).ListUnits (inline)
         0     0% 33.00%   112.65MB 19.70%  github.com/coreos/go-systemd/v22/dbus.(*Conn).ListUnitsContext
   67.51MB 11.80% 44.81%   112.51MB 19.67%  debug/elf.NewFile
   29.11MB  5.09% 49.90%   112.15MB 19.61%  github.com/coreos/go-systemd/v22/dbus.(*Conn).listUnitsInternal
         0     0% 49.90%   108.55MB 18.98%  github.com/parca-dev/parca-agent/pkg/objectfile.(*cache).ObjectFileForProcess
         0     0% 49.90%   105.55MB 18.46%  github.com/parca-dev/parca-agent/pkg/objectfile.fromProcess
         0     0% 49.90%   104.55MB 18.28%  github.com/parca-dev/parca-agent/pkg/objectfile.Open
         0     0% 49.90%   104.55MB 18.28%  github.com/parca-dev/parca-agent/pkg/objectfile.open

ROUTINE ======================== compress/flate.NewWriter in compress/flate/deflate.go
    6.17MB     9.71MB (flat, cum)  1.70% of Total
         .          .    665:// compression efficiency.
         .          .    666://
         .          .    667:// If level is in the range [-2, 9] then the error returned will be nil.
         .          .    668:// Otherwise the error returned will be non-nil.
         .          .    669:func NewWriter(w io.Writer, level int) (*Writer, error) {
    6.17MB     6.17MB    670:	var dw Writer
         .     3.54MB    671:	if err := dw.d.init(w, level); err != nil {
         .          .    672:		return nil, err
         .          .    673:	}
         .          .    674:	return &dw, nil
         .          .    675:}
         .          .    676:
ROUTINE ======================== compress/gzip.(*Writer).Write in compress/gzip/gzip.go
         0     9.71MB (flat, cum)  1.70% of Total
         .          .    186:			if z.err != nil {
         .          .    187:				return 0, z.err
         .          .    188:			}
         .          .    189:		}
         .          .    190:		if z.compressor == nil {
         .     9.71MB    191:			z.compressor, _ = flate.NewWriter(z.w, z.level)
         .          .    192:		}
         .          .    193:	}
         .          .    194:	z.size += uint32(len(p))
         .          .    195:	z.digest = crc32.Update(z.digest, crc32.IEEETable, p)
         .          .    196:	n, z.err = z.compressor.Write(p)
```

</details>

Reply:
Creating a gzip file without a footer wasn't a problem to read the original "hello world" string.

<details>

```go
package main

import (
	"bytes"
	"compress/gzip"
	"fmt"
	"io"
	"log"
	"os"
)

func main() {
	f, err := os.CreateTemp("", "")
	if err != nil {
		log.Fatal(err)
	}

	zw := gzip.NewWriter(nil)
	buf := bytes.NewBuffer(nil)
	zw.Reset(buf)

	if _, err = zw.Write([]byte("hello world")); err != nil {
		log.Fatal(err)
	}
	zw.Flush()

	f.Write(buf.Bytes())
	f.Close()

	f, err = os.Open(f.Name())
	if err != nil {
		log.Fatal(err)
	}

	zr, err := gzip.NewReader(f)
	if err != nil {
		log.Fatal(err)
	}
	got := &bytes.Buffer{}
	io.Copy(got, zr)
	fmt.Println(got)
}
```

</details>